provider: avoid CSI secret name collisions

leelavg opened the pull request:
provider sends the ceph client secret name to onboarded consumer as gRPC response, however if both provider & client operators running in same namespaces the secrets referred will collide. So, we'll send different secret names with same content to be created by client.

leelavg commented:
Obsoletes #2537

leelavg commented:
Testing:
```
# k get secret | grep '^ceph-client'
ceph-client-node-0c4ae5e2928d6f00ad8cc6827c670082               Opaque                                2      8m8s
ceph-client-node-e5edf8a0d7c81df26b30e879b47918e4               Opaque                                2      9m30s
ceph-client-provisioner-0c4ae5e2928d6f00ad8cc6827c670082        Opaque                                2      8m8s
ceph-client-provisioner-e5edf8a0d7c81df26b30e879b47918e4        Opaque                                2      9m30s

# k get sc -oyaml | rg 'node|provisioner'
  provisioner: ebs.csi.aws.com
  provisioner: ebs.csi.aws.com
    csi.storage.k8s.io/controller-expand-secret-name: ceph-client-provisioner-e5edf8a0d7c81df26b30e879b47918e4
    csi.storage.k8s.io/node-stage-secret-name: ceph-client-node-e5edf8a0d7c81df26b30e879b47918e4
    csi.storage.k8s.io/node-stage-secret-namespace: openshift-storage
    csi.storage.k8s.io/provisioner-secret-name: ceph-client-provisioner-e5edf8a0d7c81df26b30e879b47918e4
    csi.storage.k8s.io/provisioner-secret-namespace: openshift-storage
  provisioner: openshift-storage.rbd.csi.ceph.com
    csi.storage.k8s.io/controller-expand-secret-name: ceph-client-provisioner-0c4ae5e2928d6f00ad8cc6827c670082
    csi.storage.k8s.io/node-stage-secret-name: ceph-client-node-0c4ae5e2928d6f00ad8cc6827c670082
    csi.storage.k8s.io/node-stage-secret-namespace: openshift-storage
    csi.storage.k8s.io/provisioner-secret-name: ceph-client-provisioner-0c4ae5e2928d6f00ad8cc6827c670082
    csi.storage.k8s.io/provisioner-secret-namespace: openshift-storage
  provisioner: openshift-storage.cephfs.csi.ceph.com
  provisioner: openshift-storage.noobaa.io/obc
  
# k get vsclass -oyaml | rg provisioner
    csi.storage.k8s.io/snapshotter-secret-name: ceph-client-provisioner-e5edf8a0d7c81df26b30e879b47918e4
    csi.storage.k8s.io/snapshotter-secret-name: ceph-client-provisioner-0c4ae5e2928d6f00ad8cc6827c670082
    
# k get po,pvc,vs -ndefault -lpart-of=test
NAME                      READY   STATUS    RESTARTS   AGE
pod/app-block-1           1/1     Running   0          8m28s
pod/app-block-clone-1     1/1     Running   0          8m27s
pod/app-block-restore-1   1/1     Running   0          8m27s
pod/app-file-1            1/1     Running   0          8m27s
pod/app-file-clone-1      1/1     Running   0          8m25s
pod/app-file-restore-1    1/1     Running   0          8m26s

NAME                                         STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS                  AGE
persistentvolumeclaim/ceph-block-1           Bound    pvc-adca5da4-0bd6-4e8f-9373-b8ca83281808   100Mi      RWO            ocs-storagecluster-ceph-rbd   8m28s
persistentvolumeclaim/ceph-block-clone-1     Bound    pvc-dc111eff-a9f8-4125-9a5d-edc119aa15e2   100Mi      RWO            ocs-storagecluster-ceph-rbd   8m27s
persistentvolumeclaim/ceph-block-restore-1   Bound    pvc-29c9c88c-8f4e-498f-9acc-eeaf2d5d72d3   100Mi      RWO            ocs-storagecluster-ceph-rbd   8m28s
persistentvolumeclaim/ceph-file-1            Bound    pvc-638a52ae-d269-42d9-8835-247415aa1ee8   100Mi      RWO            ocs-storagecluster-cephfs     8m27s
persistentvolumeclaim/ceph-file-clone-1      Bound    pvc-48504879-8e4c-4784-ac06-24cdbff37624   100Mi      RWO            ocs-storagecluster-cephfs     8m26s
persistentvolumeclaim/ceph-file-restore-1    Bound    pvc-0105c228-e7b9-44b3-ac9e-7f6ce1222c16   100Mi      RWO            ocs-storagecluster-cephfs     8m26s

NAME                                                       READYTOUSE   SOURCEPVC      SOURCESNAPSHOTCONTENT   RESTORESIZE   SNAPSHOTCLASS                 SNAPSHOTCONTENT                                    CREATIONTIME   AGE
volumesnapshot.snapshot.storage.k8s.io/ceph-block-snap-1   true         ceph-block-1                           100Mi         ocs-storagecluster-ceph-rbd   snapcontent-d8375fd3-3871-4261-b187-0061ada352e9   8m26s          8m28s
volumesnapshot.snapshot.storage.k8s.io/ceph-file-snap-1    true         ceph-file-1                            100Mi         ocs-storagecluster-cephfs     snapcontent-cbab49df-c00b-4036-aa83-4fdea7f36946   8m26s          8m26s
```

nb-ohad commented on services/provider/server/server.go:
```diff
@@ -550,16 +554,25 @@ func (s *OCSProviderServer) GetStorageClaimConfig(ctx context.Context, req *pb.S
 
-	var storageClassName string
-	storageClassData := map[string]string{}
+	storageRequestHash := getStorageRequestHash(req.StorageConsumerUUID, req.StorageClaimName)
+	provisionerSecretName := storageClaimCephCsiSecretName("provisioner", storageRequestHash)
+	nodeSecretName := storageClaimCephCsiSecretName("node", storageRequestHash)
 
+	storageClassData := map[string]string{
+		"csi.storage.k8s.io/provisioner-secret-name":       provisionerSecretName,
+		"csi.storage.k8s.io/node-stage-secret-name":        nodeSecretName,
+		"csi.storage.k8s.io/controller-expand-secret-name": provisionerSecretName,
+	}
```
Why are you assuming these secrets are in the reply and precalculate them? You should only invoke these based on reply content. And you are already doing so inside the switch statement

leelavg replied:
Actually, this was also my first comment to Jose's PR and later realised this block will only be hit after the StorageRequest phase is READY by which time it is guaranteed the secrets will exist.

leelavg replied:
changed code to explicitly check for the secret existence. done.

nb-ohad commented on services/provider/server/storagerequest.go:
```diff
@@ -32,45 +32,51 @@ func newStorageRequestManager(cl client.Client, namespace string) (*storageReque
+	s := struct {
 		StorageConsumerUUID string `json:"storageConsumerUUID"`
-		StorageRequestName  string `json:"storageRequestName"`
+		// name in json tag isn't changed for backward compatibility or else md5 sum of json marshal will change
+		StorageClaimName string `json:"storageRequestName"`
```
The secret names are already changing so why do we care about changing this?

leelavg replied:
This isn't for secret name but for the name of the storagerequest itself. Earlier it was `storageclassrequest-hash123` now it'll be `storagerquest-hash123` it needs to be like this for migration.

leelavg replied:
changed the tag as well, done. Since underlying ceph resources names itself will be changed, having this backward compatible doesn't server any purpose, will raise new PR for a consistent naming of cbprns and cfssvg.

nb-ohad commented on services/provider/server/server.go:
```diff
@@ -590,12 +593,12 @@ func (s *OCSProviderServer) GetStorageClaimConfig(ctx context.Context, req *pb.S
 			storageClassData["radosnamespace"] = cephRes.Name
 
-			nodeCephClientSecret, _, err := s.getCephClientInformation(ctx, cephRes.CephClients["node"])
+			_, _, err := s.getCephClientInformation(ctx, cephRes.CephClients["node"])
```
Why are you calling this if you are not interested in the return values?

nb-ohad commented on services/provider/server/server.go:
```diff
@@ -590,12 +593,12 @@ func (s *OCSProviderServer) GetStorageClaimConfig(ctx context.Context, req *pb.S
 			}
 
-			provisionerCephClientSecret, _, err := s.getCephClientInformation(ctx, cephRes.CephClients["provisioner"])
+			_, _, err = s.getCephClientInformation(ctx, cephRes.CephClients["provisioner"])
```
Why are you calling this if you are not interested in the return values?

nb-ohad commented on services/provider/server/server.go:
```diff
@@ -633,24 +638,26 @@ func (s *OCSProviderServer) GetStorageClaimConfig(ctx context.Context, req *pb.S
-			storageClassData["csi.storage.k8s.io/provisioner-secret-name"] = provisionerCephClientSecret
-			storageClassData["csi.storage.k8s.io/node-stage-secret-name"] = nodeCephClientSecret
-			storageClassData["csi.storage.k8s.io/controller-expand-secret-name"] = provisionerCephClientSecret
+			storageClassData["csi.storage.k8s.io/provisioner-secret-name"] = provisionerSecretName
+			storageClassData["csi.storage.k8s.io/node-stage-secret-name"] = nodeSecretName
+			storageClassData["csi.storage.k8s.io/controller-expand-secret-name"] = provisionerSecretName
```
I don't see you appending the storage class. I can see something happens after the switch but the does not make sense. You should create it inside the case for that storage class.

leelavg replied:
done.

Madhu-1 commented on services/provider/server/server.go:
```diff
@@ -633,24 +632,23 @@ func (s *OCSProviderServer) GetStorageClaimConfig(ctx context.Context, req *pb.S
+				Data: mustMarshal(cephfsStorageClassData),
+			})
 
 			extR = append(extR, &pb.ExternalResource{
```
this section might also be not required, please check it.

leelavg replied:
as discussed offline, this is redundant data being sent which can be removed at a later point.

Madhu-1 reviewed (commented):
LGTM

leelavg commented:
@iamniting could you pls approve it as I've already addressed all comments and Madhu also approved above.

leelavg commented:
successful run after addressing all the review comments are as below, CI can be overriden.
```
[1782273049018503168](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/red-hat-storage_ocs-operator/2571/pull-ci-red-hat-storage-ocs-operator-main-ocs-operator-bundle-e2e-aws/1782273049018503168)	#[2571](https://github.com/red-hat-storage/ocs-operator/pull/2571)
([834831d](https://github.com/red-hat-storage/ocs-operator/pull/2571/commits/834831dd957568afd57a9dfb5fd20ead25be36aa)) by [leelavg](https://github.com/leelavg)	
10:29:46
1h11m8s	SUCCESS
```

iamniting commented:
/override ci/prow/ocs-operator-bundle-e2e-aws
